# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [5]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    
    # Convert characters to ids 
    english_vocab_ids =[[source_vocab_to_int[vocab] for vocab in sentence.split()] for sentence in source_text.split('\n')]
    french_vocab_ids= [[target_vocab_to_int[vocab] for vocab in sentence.split()] + [target_vocab_to_int['<EOS>']] for sentence in target_text.split('\n')]
    
    return english_vocab_ids, french_vocab_ids

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


/Users/macpro/anaconda/envs/imgcls/lib/python3.6/site-packages/ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [36]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input = tf.placeholder(tf.int32, shape = [None, None], name = "input")
    targets = tf.placeholder(tf.int32, shape = [None, None], name = "targets")
    learning_rate= tf.placeholder(tf.float32, name = "learning_rate")
    keep_prob = tf.placeholder(tf.float32, name = "keep_prob")
    target_sequence_length = tf.placeholder(tf.int32, shape = [None], name = "target_sequence_length")
    max_target_len = tf.reduce_max(target_sequence_length)
    source_sequence_length = tf.placeholder(tf.int32, shape = [None], name = "source_sequence_length")
    return input, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [37]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0,0],[batch_size, -1],[1,1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']),ending],1)
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [50]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    ##RNN Cell
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                          initializer = tf.random_uniform_initializer(-0.1,0.1, seed=2))
        return enc_cell
    
    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, sequence_length = source_sequence_length,
                                             dtype= tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [45]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                        sequence_length=target_sequence_length,
                                                        time_major=False)
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                        training_helper,
                                                        encoder_state,
                                                        output_layer)
    training_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                   impute_finished=True,
                                                                   maximum_iterations=max_summary_length)
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [46]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype = tf.int32),[batch_size], name = 'start_tokens')
    
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)
    
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, inference_helper, encoder_state, output_layer)
    
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder, impute_finished=True,
                                                                 maximum_iterations = max_target_sequence_length)[0]
    return inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [47]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    def make_cell(rnn_size):
        return tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean=0.0, stddev=0.1))
    
    with tf.variable_scope("decode"):
        training_logits = decoding_layer_train(encoder_state, dec_cell, dec_embed_input,
                                             target_sequence_length, max_target_sequence_length,
                                             output_layer, keep_prob)
        
        
    with tf.variable_scope("decode", reuse=True):
        inference_logits = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, target_vocab_to_int['<GO>'],
                                             target_vocab_to_int['<EOS>'], max_target_sequence_length,
                                             target_vocab_size, output_layer, batch_size, keep_prob)
    
    return training_logits, inference_logits



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [52]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)

    """
#     enc_embed_input = tf.contrib.layers.embed_sequence(
#         input_data, source_vocab_size, enc_embedding_size)
    
    _,enc_state = encoding_layer(
        input_data, rnn_size, num_layers, keep_prob, source_sequence_length, source_vocab_size, enc_embedding_size)
    
#     proc_target_data = process_decoding_input(target_data, target_vocab_to_int, batch_size)
#     dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, dec_embedding_size]))
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    train_logits, refer_logits = decoding_layer(dec_input, enc_state, target_sequence_length, 
                                                max_target_sentence_length, rnn_size, num_layers,
                                                target_vocab_to_int, target_vocab_size, batch_size, 
                                                keep_prob, dec_embedding_size)
    
    return train_logits, refer_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [53]:
# Number of Epochs
epochs = 6
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 30
decoding_embedding_size = 30

# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.8
display_step = 3

### Build the Graph
Build the graph using the neural network you implemented.

In [54]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [55]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [56]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    3/538 - Train Accuracy: 0.0746, Validation Accuracy: 0.0639, Loss: 5.0531
Epoch   0 Batch    6/538 - Train Accuracy: 0.3097, Validation Accuracy: 0.3517, Loss: 3.5929
Epoch   0 Batch    9/538 - Train Accuracy: 0.2840, Validation Accuracy: 0.3519, Loss: 3.4218
Epoch   0 Batch   12/538 - Train Accuracy: 0.2969, Validation Accuracy: 0.3693, Loss: 3.2769
Epoch   0 Batch   15/538 - Train Accuracy: 0.3462, Validation Accuracy: 0.3768, Loss: 3.0064
Epoch   0 Batch   18/538 - Train Accuracy: 0.3246, Validation Accuracy: 0.3915, Loss: 3.0861
Epoch   0 Batch   21/538 - Train Accuracy: 0.2895, Validation Accuracy: 0.3949, Loss: 3.0488
Epoch   0 Batch   24/538 - Train Accuracy: 0.3359, Validation Accuracy: 0.3716, Loss: 2.9874
Epoch   0 Batch   27/538 - Train Accuracy: 0.3387, Validation Accuracy: 0.3993, Loss: 2.9356
Epoch   0 Batch   30/538 - Train Accuracy: 0.3344, Validation Accuracy: 0.4055, Loss: 2.8301
Epoch   0 Batch   33/538 - Train Accuracy: 0.3696, Validation Accuracy

Epoch   0 Batch  270/538 - Train Accuracy: 0.6488, Validation Accuracy: 0.6460, Loss: 0.5301
Epoch   0 Batch  273/538 - Train Accuracy: 0.6498, Validation Accuracy: 0.6555, Loss: 0.5387
Epoch   0 Batch  276/538 - Train Accuracy: 0.6477, Validation Accuracy: 0.6452, Loss: 0.5374
Epoch   0 Batch  279/538 - Train Accuracy: 0.6488, Validation Accuracy: 0.6147, Loss: 0.5213
Epoch   0 Batch  282/538 - Train Accuracy: 0.6602, Validation Accuracy: 0.6422, Loss: 0.5154
Epoch   0 Batch  285/538 - Train Accuracy: 0.6408, Validation Accuracy: 0.6602, Loss: 0.4892
Epoch   0 Batch  288/538 - Train Accuracy: 0.6141, Validation Accuracy: 0.6406, Loss: 0.5262
Epoch   0 Batch  291/538 - Train Accuracy: 0.6685, Validation Accuracy: 0.6523, Loss: 0.5234
Epoch   0 Batch  294/538 - Train Accuracy: 0.6271, Validation Accuracy: 0.6399, Loss: 0.5457
Epoch   0 Batch  297/538 - Train Accuracy: 0.6336, Validation Accuracy: 0.6452, Loss: 0.5170
Epoch   0 Batch  300/538 - Train Accuracy: 0.6659, Validation Accuracy

Epoch   1 Batch    3/538 - Train Accuracy: 0.8295, Validation Accuracy: 0.7852, Loss: 0.2216
Epoch   1 Batch    6/538 - Train Accuracy: 0.8064, Validation Accuracy: 0.7710, Loss: 0.1964
Epoch   1 Batch    9/538 - Train Accuracy: 0.8096, Validation Accuracy: 0.7708, Loss: 0.2063
Epoch   1 Batch   12/538 - Train Accuracy: 0.7834, Validation Accuracy: 0.7852, Loss: 0.2025
Epoch   1 Batch   15/538 - Train Accuracy: 0.8317, Validation Accuracy: 0.7729, Loss: 0.1888
Epoch   1 Batch   18/538 - Train Accuracy: 0.8414, Validation Accuracy: 0.7889, Loss: 0.2027
Epoch   1 Batch   21/538 - Train Accuracy: 0.8150, Validation Accuracy: 0.8029, Loss: 0.1824
Epoch   1 Batch   24/538 - Train Accuracy: 0.8086, Validation Accuracy: 0.7901, Loss: 0.1940
Epoch   1 Batch   27/538 - Train Accuracy: 0.8025, Validation Accuracy: 0.7894, Loss: 0.1776
Epoch   1 Batch   30/538 - Train Accuracy: 0.7957, Validation Accuracy: 0.7914, Loss: 0.2024
Epoch   1 Batch   33/538 - Train Accuracy: 0.8140, Validation Accuracy

Epoch   1 Batch  270/538 - Train Accuracy: 0.8893, Validation Accuracy: 0.8558, Loss: 0.0710
Epoch   1 Batch  273/538 - Train Accuracy: 0.9055, Validation Accuracy: 0.8739, Loss: 0.0783
Epoch   1 Batch  276/538 - Train Accuracy: 0.8945, Validation Accuracy: 0.8572, Loss: 0.0738
Epoch   1 Batch  279/538 - Train Accuracy: 0.8775, Validation Accuracy: 0.8821, Loss: 0.0689
Epoch   1 Batch  282/538 - Train Accuracy: 0.8884, Validation Accuracy: 0.8954, Loss: 0.0702
Epoch   1 Batch  285/538 - Train Accuracy: 0.8906, Validation Accuracy: 0.8951, Loss: 0.0595
Epoch   1 Batch  288/538 - Train Accuracy: 0.9223, Validation Accuracy: 0.8990, Loss: 0.0623
Epoch   1 Batch  291/538 - Train Accuracy: 0.9005, Validation Accuracy: 0.8857, Loss: 0.0627
Epoch   1 Batch  294/538 - Train Accuracy: 0.9034, Validation Accuracy: 0.8738, Loss: 0.0718
Epoch   1 Batch  297/538 - Train Accuracy: 0.9260, Validation Accuracy: 0.8936, Loss: 0.0609
Epoch   1 Batch  300/538 - Train Accuracy: 0.9126, Validation Accuracy

Epoch   2 Batch    3/538 - Train Accuracy: 0.9244, Validation Accuracy: 0.9125, Loss: 0.0423
Epoch   2 Batch    6/538 - Train Accuracy: 0.9211, Validation Accuracy: 0.9144, Loss: 0.0453
Epoch   2 Batch    9/538 - Train Accuracy: 0.9021, Validation Accuracy: 0.9261, Loss: 0.0457
Epoch   2 Batch   12/538 - Train Accuracy: 0.9232, Validation Accuracy: 0.8910, Loss: 0.0467
Epoch   2 Batch   15/538 - Train Accuracy: 0.9234, Validation Accuracy: 0.8968, Loss: 0.0422
Epoch   2 Batch   18/538 - Train Accuracy: 0.9396, Validation Accuracy: 0.8942, Loss: 0.0543
Epoch   2 Batch   21/538 - Train Accuracy: 0.9391, Validation Accuracy: 0.9237, Loss: 0.0370
Epoch   2 Batch   24/538 - Train Accuracy: 0.9189, Validation Accuracy: 0.9173, Loss: 0.0470
Epoch   2 Batch   27/538 - Train Accuracy: 0.9291, Validation Accuracy: 0.9112, Loss: 0.0418
Epoch   2 Batch   30/538 - Train Accuracy: 0.9074, Validation Accuracy: 0.9146, Loss: 0.0572
Epoch   2 Batch   33/538 - Train Accuracy: 0.9351, Validation Accuracy

Epoch   2 Batch  270/538 - Train Accuracy: 0.9449, Validation Accuracy: 0.9224, Loss: 0.0361
Epoch   2 Batch  273/538 - Train Accuracy: 0.9445, Validation Accuracy: 0.9206, Loss: 0.0406
Epoch   2 Batch  276/538 - Train Accuracy: 0.9391, Validation Accuracy: 0.9185, Loss: 0.0442
Epoch   2 Batch  279/538 - Train Accuracy: 0.9492, Validation Accuracy: 0.9398, Loss: 0.0327
Epoch   2 Batch  282/538 - Train Accuracy: 0.9522, Validation Accuracy: 0.9405, Loss: 0.0337
Epoch   2 Batch  285/538 - Train Accuracy: 0.9555, Validation Accuracy: 0.9426, Loss: 0.0279
Epoch   2 Batch  288/538 - Train Accuracy: 0.9502, Validation Accuracy: 0.9332, Loss: 0.0314
Epoch   2 Batch  291/538 - Train Accuracy: 0.9518, Validation Accuracy: 0.9348, Loss: 0.0328
Epoch   2 Batch  294/538 - Train Accuracy: 0.9379, Validation Accuracy: 0.9496, Loss: 0.0365
Epoch   2 Batch  297/538 - Train Accuracy: 0.9625, Validation Accuracy: 0.9521, Loss: 0.0315
Epoch   2 Batch  300/538 - Train Accuracy: 0.9462, Validation Accuracy

Epoch   3 Batch    3/538 - Train Accuracy: 0.9750, Validation Accuracy: 0.9462, Loss: 0.0236
Epoch   3 Batch    6/538 - Train Accuracy: 0.9580, Validation Accuracy: 0.9398, Loss: 0.0277
Epoch   3 Batch    9/538 - Train Accuracy: 0.9326, Validation Accuracy: 0.9439, Loss: 0.0310
Epoch   3 Batch   12/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9547, Loss: 0.0301
Epoch   3 Batch   15/538 - Train Accuracy: 0.9554, Validation Accuracy: 0.9462, Loss: 0.0256
Epoch   3 Batch   18/538 - Train Accuracy: 0.9590, Validation Accuracy: 0.9487, Loss: 0.0370
Epoch   3 Batch   21/538 - Train Accuracy: 0.9706, Validation Accuracy: 0.9531, Loss: 0.0195
Epoch   3 Batch   24/538 - Train Accuracy: 0.9648, Validation Accuracy: 0.9490, Loss: 0.0288
Epoch   3 Batch   27/538 - Train Accuracy: 0.9672, Validation Accuracy: 0.9387, Loss: 0.0235
Epoch   3 Batch   30/538 - Train Accuracy: 0.9527, Validation Accuracy: 0.9501, Loss: 0.0314
Epoch   3 Batch   33/538 - Train Accuracy: 0.9650, Validation Accuracy

Epoch   3 Batch  270/538 - Train Accuracy: 0.9615, Validation Accuracy: 0.9419, Loss: 0.0243
Epoch   3 Batch  273/538 - Train Accuracy: 0.9479, Validation Accuracy: 0.9437, Loss: 0.0305
Epoch   3 Batch  276/538 - Train Accuracy: 0.9559, Validation Accuracy: 0.9492, Loss: 0.0322
Epoch   3 Batch  279/538 - Train Accuracy: 0.9680, Validation Accuracy: 0.9466, Loss: 0.0234
Epoch   3 Batch  282/538 - Train Accuracy: 0.9587, Validation Accuracy: 0.9490, Loss: 0.0270
Epoch   3 Batch  285/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9551, Loss: 0.0227
Epoch   3 Batch  288/538 - Train Accuracy: 0.9572, Validation Accuracy: 0.9476, Loss: 0.0241
Epoch   3 Batch  291/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9448, Loss: 0.0232
Epoch   3 Batch  294/538 - Train Accuracy: 0.9562, Validation Accuracy: 0.9551, Loss: 0.0239
Epoch   3 Batch  297/538 - Train Accuracy: 0.9797, Validation Accuracy: 0.9648, Loss: 0.0222
Epoch   3 Batch  300/538 - Train Accuracy: 0.9658, Validation Accuracy

Epoch   4 Batch    3/538 - Train Accuracy: 0.9703, Validation Accuracy: 0.9444, Loss: 0.0177
Epoch   4 Batch    6/538 - Train Accuracy: 0.9691, Validation Accuracy: 0.9393, Loss: 0.0195
Epoch   4 Batch    9/538 - Train Accuracy: 0.9480, Validation Accuracy: 0.9469, Loss: 0.0245
Epoch   4 Batch   12/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9510, Loss: 0.0218
Epoch   4 Batch   15/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9636, Loss: 0.0201
Epoch   4 Batch   18/538 - Train Accuracy: 0.9631, Validation Accuracy: 0.9632, Loss: 0.0282
Epoch   4 Batch   21/538 - Train Accuracy: 0.9870, Validation Accuracy: 0.9585, Loss: 0.0114
Epoch   4 Batch   24/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9517, Loss: 0.0236
Epoch   4 Batch   27/538 - Train Accuracy: 0.9773, Validation Accuracy: 0.9460, Loss: 0.0195
Epoch   4 Batch   30/538 - Train Accuracy: 0.9572, Validation Accuracy: 0.9542, Loss: 0.0258
Epoch   4 Batch   33/538 - Train Accuracy: 0.9581, Validation Accuracy

Epoch   4 Batch  270/538 - Train Accuracy: 0.9729, Validation Accuracy: 0.9480, Loss: 0.0168
Epoch   4 Batch  273/538 - Train Accuracy: 0.9555, Validation Accuracy: 0.9338, Loss: 0.0238
Epoch   4 Batch  276/538 - Train Accuracy: 0.9637, Validation Accuracy: 0.9453, Loss: 0.0297
Epoch   4 Batch  279/538 - Train Accuracy: 0.9684, Validation Accuracy: 0.9544, Loss: 0.0197
Epoch   4 Batch  282/538 - Train Accuracy: 0.9630, Validation Accuracy: 0.9542, Loss: 0.0229
Epoch   4 Batch  285/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9560, Loss: 0.0185
Epoch   4 Batch  288/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9544, Loss: 0.0211
Epoch   4 Batch  291/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9537, Loss: 0.0211
Epoch   4 Batch  294/538 - Train Accuracy: 0.9677, Validation Accuracy: 0.9641, Loss: 0.0208
Epoch   4 Batch  297/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9654, Loss: 0.0213
Epoch   4 Batch  300/538 - Train Accuracy: 0.9667, Validation Accuracy

Epoch   5 Batch    3/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9508, Loss: 0.0160
Epoch   5 Batch    6/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9561, Loss: 0.0244
Epoch   5 Batch    9/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9600, Loss: 0.0276
Epoch   5 Batch   12/538 - Train Accuracy: 0.9725, Validation Accuracy: 0.9565, Loss: 0.0250
Epoch   5 Batch   15/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9600, Loss: 0.0197
Epoch   5 Batch   18/538 - Train Accuracy: 0.9635, Validation Accuracy: 0.9613, Loss: 0.0261
Epoch   5 Batch   21/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9558, Loss: 0.0140
Epoch   5 Batch   24/538 - Train Accuracy: 0.9723, Validation Accuracy: 0.9466, Loss: 0.0247
Epoch   5 Batch   27/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9576, Loss: 0.0164
Epoch   5 Batch   30/538 - Train Accuracy: 0.9586, Validation Accuracy: 0.9537, Loss: 0.0254
Epoch   5 Batch   33/538 - Train Accuracy: 0.9764, Validation Accuracy

Epoch   5 Batch  270/538 - Train Accuracy: 0.9725, Validation Accuracy: 0.9535, Loss: 0.0187
Epoch   5 Batch  273/538 - Train Accuracy: 0.9660, Validation Accuracy: 0.9505, Loss: 0.0264
Epoch   5 Batch  276/538 - Train Accuracy: 0.9672, Validation Accuracy: 0.9492, Loss: 0.0277
Epoch   5 Batch  279/538 - Train Accuracy: 0.9660, Validation Accuracy: 0.9542, Loss: 0.0188
Epoch   5 Batch  282/538 - Train Accuracy: 0.9712, Validation Accuracy: 0.9588, Loss: 0.0230
Epoch   5 Batch  285/538 - Train Accuracy: 0.9727, Validation Accuracy: 0.9521, Loss: 0.0175
Epoch   5 Batch  288/538 - Train Accuracy: 0.9566, Validation Accuracy: 0.9540, Loss: 0.0245
Epoch   5 Batch  291/538 - Train Accuracy: 0.9794, Validation Accuracy: 0.9599, Loss: 0.0189
Epoch   5 Batch  294/538 - Train Accuracy: 0.9626, Validation Accuracy: 0.9622, Loss: 0.0223
Epoch   5 Batch  297/538 - Train Accuracy: 0.9744, Validation Accuracy: 0.9585, Loss: 0.0210
Epoch   5 Batch  300/538 - Train Accuracy: 0.9732, Validation Accuracy

Model Trained and Saved


### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [57]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [58]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [61]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    unk = vocab_to_int['<UNK>']
    
    return [vocab_to_int.get(w, unk) for w in sentence.lower().split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [62]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [39, 182, 46, 76, 131, 190, 199]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [181, 183, 303, 204, 344, 193, 252, 82, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.